numpy/core/_internal.py

```diff
@@ -355,15 +355,38 @@ def _view_is_safe(oldtype, newtype):
         If the new type is incompatible with the old type.
 
     """
+    from numpy.lib.type_check import find_dtype_offsets
 
     # if the types are equivalent, there is no problem.
     # for example: dtype((np.record, 'i4,i4')) == dtype((np.void, 'i4,i4'))
     if oldtype == newtype:
         return
 
-    if newtype.hasobject or oldtype.hasobject:
-        raise TypeError("Cannot change data-type for object array.")
-    return
+    # no object members is fine
+    if not newtype.hasobject and not oldtype.hasobject:
+        return
```

Comment:
What about this:

First, I think we only need some relatively minor updates so that the `hasobject` field is guaranteed to be `True` if the dtype has objects *including hidden objects*. Thus, whenever taking a view of an array with objects, the view dtype will have `hasobject=true` even if the objects are hidden. 

This is actually often the case as implemented in #6053 - with  `c = np.zeros(4, dtype="p,O,p")[['f0','f2']]`, I see that `c.dtype.hasobject` is `True`. However, I'm pretty sure I saw other cases where I can get `hasobject` to be false if there are hidden objects. Those would need to be fixed.

Then, this PR also needs a modification: The view should only be allowed if *all* the fields (not just object fields) are present at the right position in the view. This way `b`.view('O,p,O') is disallowed since the p is missing in `a`. (And if hasobject=False for both inputs, the view is allowed).

I think that restriction on this PR still allows most of the use-cases you've pointed out.